### PR TITLE
middleware/consul: new consul middleware

### PR DESCRIPTION
**Pull request by VioletRainbows**
#712  - New Consul middleware

#### Features not implemented
- Consul Service Discovery

#### Features implemented
- All features in etcd should be functional in Consul

#### Dev
The main file is `consul.go`. All other files are the same as `etcd` minus a bit of renaming.

I also removed some reference to Skydns in `README.md` as to not confuse peeps into thinking Skydns supports Consul.

### CODE REVIEW

File: middleware/consul/README.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
+# Consul
+
+*Consul* enables reading zone data from an Consul instance. The data in Consul has to be encoded as
+a [message](https://github.com/skynetservices/skydns/blob/2fcff74cdc9f9a7dd64189a447ef27ac354b725f/msg/service.go#L26)
```
Comment (miekg) [inline middleware/consul/README.md]: Note I've pulled that definition into CoreDNS as well: middleware/etcd/msg

File: middleware/consul/consul.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	"github.com/coredns/coredns/middleware/proxy"
+	"github.com/coredns/coredns/request"
+
+	consulapi "github.com/hashicorp/consul/api"
+	"github.com/miekg/dns"
+)
+
+var errNoItems = errors.New("no items found")
+
+// Consul is a middleware talks to an etcd cluster.
```
Comment (miekg) [inline middleware/consul/consul.go]: etcd? really?

**Review by miekg — COMMENTED** [review]
Awesome to submit this middleware!

Gave it a very early review (more this weekend). There is still a bunch of references to etcd in the comments mostly.

File: middleware/consul/README.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+* `debug` allows for debug queries. Prefix the name with `o-o.debug.` to retrieve extra information in the
+  additional section of the reply in the form of TXT records.
+
+## Examples
+
+This is the default CoreDNS setup, with everying specified in full:
+
+~~~
+.:53 {
+    consul CoreDNS.local {
```
Comment (miekg) [inline middleware/consul/README.md:56]: nit: just lowercase 'CoreDNS' here as it is a domain name, not the name CoreDNS

File: middleware/consul/README.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
+# Consul
+
+*Consul* enables reading zone data from an Consul instance. The data in Consul has to be encoded as
```
Comment (miekg) [inline middleware/consul/README.md:3]: What actually is consul? Just etcd?

File: middleware/consul/setup.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	// config.Scheme is handled in NewClient()
+	client, err := consulapi.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func newHTTPSTransport(cc *tls.Config) *http.Transport {
```
Comment (miekg) [inline middleware/consul/setup.go:147]: We should extract these out and put them in a `middleware/pkg` package

File: middleware/consul/README.md (hunk 4 above)
Comment (miekg) [reply middleware/consul/README.md:3]: Is consul etcd3 or etcd2?

**Review by miekg — COMMENTED** [review]
Thanks for this PR, consul is a nice addition to CoreDNS and we should add it.
Right now it's a lot of copy and paste from etcd (skydns) backend. Iff consul is etcd3 we should probably drop our etcd2 middleware and make that etcd3 and then extract *all* the similarities and make consul use that (or vice versa).

WDTY?

**Comment by VioletRainbows** [comment]
I'm not sure I understand fully, but here is my take at it:

If you are saying that we should reduce code duplication as much as possible between the etcd and consul middleware, then I do agree.

Would it be preferable to refactor the code in a way so that only the interface communicating with etcd{v2,v3} and consul are needed?
Quick example:
```go
// need better name than foo
type foo interface {
   Records(name string, exact bool) (records []msg.Service, err error)
}
```

We would only need to implement the `Records` function for each backend. Also need figure out the initialization and/or `Corefile` and documentation.

_FYI, I'm also trying to add etcdv3 support._

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] middleware/co..." ]
>I'm not sure I understand fully, but here is my take at it:
>
>If you are saying that we should reduce code duplication as much as possible between the etcd and consul middleware, then I do agree.

yep.

>Would it be preferable to refactor the code in a way so that only the interface communicating with etcd{v2,v3} and consul are needed?
>Quick example:
>```go
>// need better name than foo
>type foo interface {
>    func (f *Foo) Records(name string, exact bool) (records []msg.Service, err error)
>}
>```
>
>We would only need to implement the `Records` function for each backend. Also need figure out the initialization and/or `Corefile`.

Yes, but doesn't
https://github.com/coredns/coredns/blob/master/middleware/backend.go already
give you must of this? Maybe we can extract more?

>_FYI, I'm also trying to add etcdv3 support._

NICE!

Oh, one other thing. I think we should separate the 'go dep' PR from the actual
code added. With those split we only have mechanical 'go dep' update PRs.
(Trouble is that the bloody thing then fails building in travis - hmmmm, might
be harder than I anticipated)

**Comment by VioletRainbows** [comment]
I'm not sure how to implement it.

Here is what I have so far:
### Step 1
- Rename `middleware/etcd` to something generic (`foo` in the examples)
- Modify the tests to run on a list of backends (just `etcdv2` for now)
- Separate `setup.go` between generic initialization and `etcdv2` initialization
- Move `middleware/etcd/etcd.go` to`middleware/foo/etcdv2/etcd.go`
- Strip `middleware/foo/README.md` from `etcdv2` items only
- Add `middleware/foo/etcdv2/README.md` for `etcdv2` specific docs
### Step 2
- Add `middleware/foo/consul/consul.go`, `middleware/foo/consul/setup.go` and `middleware/foo/consul/README.md`
- Add `consul` backend for tests
### Step 3
- Add `middleware/foo/etcdv3/etcd.go`, `middleware/foo/etcdv3/etcd.go` and `middleware/foo/etcdv3/README.md`
- Add `etcdv3` backend for tests

All other solutions have issues with either documentation, code duplication or tests duplication.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] middleware/co..." ]
>I'm not sure how to implement it.
>
>Here is what I have so far:
>### Step 1
>- Rename `middleware/etcd` to something generic (`foo` in the examples)
>- Modify the tests to run on a list of backends (just `etcdv2` for now)
>- Separate `setup.go` between generic initialization and `etcdv2` initialization
>- Move `middleware/etcd/etcd.go` to`middleware/foo/etcdv2/etcd.go`
>- Strip `middleware/foo/README.md` from `etcdv2` items only
>- Add `middleware/foo/etcdv2/README.md` for `etcdv2` specific docs
>### Step 2
>- Add `middleware/foo/consul/consul.go`, `middleware/foo/consul/setup.go` and `middleware/foo/consul/README.md`
>- Add `consul` backend for tests
>### Step 3
>- Add `middleware/foo/etcdv3/etcd.go`, `middleware/foo/etcdv3/etcd.go` and `middleware/foo/etcdv3/README.md`
>- Add `etcdv3` backend for tests```
>
>All other solutions have issues with either documentation, code duplication or tests duplication.

Ack, this looks sane.

One last step - remove etcdv2, don't really need that.

**Comment by johnbelamaric** [comment]
That sounds like a good plan. @chrisohaver you looked at some of this with the K8s middleware. Any comment?

**Comment by chrisohaver** [comment]
Sounds good to me.

**Comment by miekg** [comment]
Can close this too, no? In favor of the work happening on the second branch.

**Pull request closed without merging**